lib/uplink/bucket.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"storj.io/storj/pkg/metainfo/kvmetainfo"
+	"storj.io/storj/pkg/storage/streams"
+	"storj.io/storj/pkg/storj"
+)
+
+// Bucket represents operations you can perform on a bucket
+type Bucket struct {
+	storj.Bucket
+
+	metainfo   *kvmetainfo.DB
+	streams    streams.Store
+	pathCipher storj.Cipher
+}
+
+// OpenObject returns an Object handle, if authorized.
+func (b *Bucket) OpenObject(ctx context.Context, path storj.Path) (o *Object, err error) {
+	defer mon.Task()(&ctx)(&err)
+	panic("TODO")
+}
+
+// UploadOptions controls options about uploading a new Object, if authorized.
+type UploadOptions struct {
+	Metadata map[string]string
+	Expires  time.Time
+
+	// EncryptionScheme determines the object's encryption scheme. If not set, uses the Uplink default
+	EncryptionScheme *storj.EncryptionScheme
```

Comment:
We have this in `BucketConfig.Volatile`: `DataCipher` and `EncryptionBlockSize`, no? So we should remove the `EncryptionScheme` from here.

Reply:
you're right

Reply:
Have we decided for sure, then, that all objects in a bucket must use the same encryption scheme?

Reply:
oh, i'm sorry, i read @kaloyan-raev's message wrong. i thought he was saying these should be Volatile fields in this struct, not remove them entirely. i think we aren't certain about how this works yet but we do probably want to let people change some details per object within a bucket.

we can either mark this field volatile here, or we can remove it (for now)

Reply:
The `EncryptionScheme` field here is duplicate to the `DataCipher` and `EncryptionBlockSize` in `BucketConfig.Volatile`.

We have to decide if we want to configure the encryption 1) per bucket (for new uploads only); or 2) per object. But we should not have both ways.

And, I believe, we should have it the same way for the RS config too. If we have encryption config per object, we should also have the RS config per object. It does not make sense to have encryption configurable per object, but RS only per bucket (for new uploads only).

Reply:
I'm not sure I agree. It seems both reasonable and useful to have a default per bucket and also allow overriding those defaults for individual objects. Is your concern about users being confused by that?

Reply:
If we want some config to have bucket's default and still be configurable per object, then the RS config is much more important than the EncryptionScheme.

It's more likely that users would want to increase the durability of specific critical files rather than encrypting different files with different ciphers.

If we want to put this flexibility in the API now then let's do it for both RS and Encryption.

Reply:
I'm also not sure if that's really more likely, but I have no objection to adding that as an explicit option per object. Let's put both RS and encryption params in a Volatile struct for now?

Reply:
ok, deal :)